apisix/plugins/datadog.lua

```diff
@@ -0,0 +1,178 @@
+-- The ASF licenses this file to You under the Apache License, Version 2.0
+-- (the "License"); you may not use this file except in compliance with
+-- the License.  You may obtain a copy of the License at
+--
+--     http://www.apache.org/licenses/LICENSE-2.0
+--
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+
+local core = require("apisix.core")
+local plugin = require("apisix.plugin")
+local send_statsd = require("apisix.plugins.udp-logger").send_udp_data
+local fetch_info = require("apisix.plugins.prometheus.exporter").parse_info_from_ctx
+local format = string.format
+local concat = table.concat
+local tostring = tostring
+local ngx = ngx
+
+
+local plugin_name = "datadog"
+
+local schema = {
+    type = "object",
+    properties = {
+    }
+}
+
+local metadata_schema = {
+    type = "object",
+    properties = {
+        host = {type = "string"},
+        port = {type = "integer", minimum = 0},
+        namespace = {type = "string", default = "apisix.dev"},
+        sample_rate = {type = "number", default = 1, minimum = 0, maximum = 1},
+        tags = {
+            type = "array",
+            items = {type = "string"},
+            default = {"source:apisix"}
+        }
+    },
+    required = {"host", "port"}
+}
+
+local _M = {
+    version = 0.1,
+    priority = 495,
+    name = plugin_name,
+    schema = schema,
+    metadata_schema = metadata_schema,
+}
+
+function _M.check_schema(conf, schema_type)
+    if schema_type == core.schema.TYPE_METADATA then
+        return core.schema.check(metadata_schema, conf)
+    end
+    return core.schema.check(schema, conf)
+end
+
+local function generate_tag(sample_rate, tag_arr, route_id, service_id,
+        consumer_name, balancer_ip, http_status)
+    local rate, tags = "", ""
+
+    if sample_rate and sample_rate ~= 1 then
+        rate = "|@" .. tostring(sample_rate)
+    end
+
+    if tag_arr and #tag_arr > 0 then
+        tags = "|#" .. concat(tag_arr, ",")
+    end
+
+    if route_id ~= "" then
+        tags = tags .. "route_id:" .. route_id
+    end
+
+    if service_id ~= "" then
+        tags = tags .. "service_id:" .. service_id
+    end
+
+    if consumer_name ~= "" then
+        tags = tags .. "consumer_name:" .. consumer_name
+    end
+    if balancer_ip ~= "" then
+        tags = tags .. "balancer_ip:" .. balancer_ip
+    end
+    if http_status then
+        tags = tags .. "http_status:" .. http_status
+    end
+
+    if tags ~= "" and tags:sub(1, 1) ~= "|" then
+        tags = "|#" .. tags
+    end
+
+    return rate .. tags
+
+end
+
+function _M.log(conf, ctx)
+    local metadata = plugin.plugin_metadata(plugin_name)
+    if not metadata then
```

Comment:
A more appreciated way is to prove a couple of the default values when metadata is missing. See https://github.com/apache/apisix/blob/759e6663162d43e761d556b85e3061f755ea74ae/apisix/plugins/batch-requests.lua#L213